[Mailer][Mime] Refactor S/MIME encryption handling in `SMimeEncryptionListener`

Spomky opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | yes
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

It appears that the smime_encrypter introduced in #58501 is incorrect, as the email is encrypted only for the sender instead of being encrypted per recipient.
This PR introduces a new `SmimeCertificateRepositoryInterface`, responsible for retrieving recipient certificates.

An email is encrypted under the following conditions:
* A certificate is found for all recipients.
* The custom header `X-SMime-Encrypt` is present.
If either of these conditions is not met, the email is sent unencrypted.

Example of usage:

```php
<?php

namespace App\Security;

use Symfony\Component\DependencyInjection\Attribute\Autowire;
use Symfony\Component\Mailer\EventListener\SmimeCertificateRepositoryInterface;

class LocalFileCertificateRepository implements SmimeCertificateRepositoryInterface
{
    public function __construct(
        #[Autowire(param: 'kernel.project_dir')]
        private readonly string $projectDir
    ){}

    public function findCertificatePathFor(string $email): ?string
    {
        $hash = hash('sha256', strtolower(trim($email)));
        $path = sprintf('%s/storage/%s.crt', $this->projectDir, $hash);

        return file_exists($path) ? $path : null;
    }
}
```

stof commented on src/Symfony/Component/Mailer/SMimeRecipient.php:
```diff
@@ -0,0 +1,25 @@
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class SMimeRecipient extends Address
```
this requires bumping the min version of the `mime` component to be the one providing a non-final class (which is a change warranting a changelog entry in the mime component btw).

stof commented on src/Symfony/Component/Mailer/EventListener/SmimeEncryptedMessageListener.php:
```diff
@@ -34,8 +35,19 @@ public function onMessage(MessageEvent $event): void
+            }
+            $certificates[] = $recipient->certificatePath;
+        }
+        if (0 === \count($certificates)) {
```
what if some recipients are not SMimeRecipient ?

Spomky replied:
Indeed, an exception should be thrown as we cannot mix encrypted/normal emails.

stof commented on src/Symfony/Component/Mailer/SMimeRecipient.php:
```diff
@@ -0,0 +1,25 @@
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class SMimeRecipient extends Address
```
I'm not sure using a dedicated recipient class is the right way to handle that need. This works only if we have a single listener needing an extra info about the recipient (like any inheritance-based extension point).

In #59372, I suggest having a service responsible for getting the encryption key for a given email address, allowing the implementation to read a static config, to load it from the DB or a cache, or any other implementation. I see several benefits (maybe I'm missing some drawbacks though):

- this API would be easier to use for projects doing centralized management of the encryption keys of their recipients (instead of requiring to use a special way to create the recipient in each place creating an email)
- this would allow providing certificate paths using temporary files *at the time encryption needs it*) when they are stored elsewhere (in a DB for instance). The API proposed here would require building the path when creating the email, which can happen on a different server than the one actually sending the email when using the messenger integration (and serializing a string containing a path does not ensure that the server deserializing it has the content at that path)
- it would allow registering several listeners dealing with extra info for the recipient (assuming they all use this kind of architecture)

What do you think about it ?

Spomky replied:
I like the idea of a service responsible for fetching the recipient certificate.
However, how to tell the listener to encrypt or not?
As an example, you have a certificate associated to the email john.doe@acme.com.
* You send an email to that email as a unique recipient. The email is encrypted by the listener.
* You send another email the john.doe and two other recipients (you also have the certificate for those two users). In this case you specifically do not want to encrypt it for any reason. But the listener will encrypt it.

This is the reason why I introduced a subclass of Address. In this case the listener know what to do.

stof replied:
if we need to support the case of encrypting *sometimes* and not encrypting *sometimes*, I would suggest adding a way to disable signing for specific emails through a metadata header.

To me, this also reduce the risk of _forgetting_ to encrypt an email in a project meaning to encrypt them all, because it is easy to use a normal Address object in a place when creating the email.

Spomky replied:
Makes sense. I revised this PR and introduced a repository.
If one of the sender has no S/MIME certificate the message is left unchanged.

Spomky replied:
What about the header `X-SMime-Encrypt` to explicitly ask for encryption? Could be used by PGP/MIME (`X-PgpMime-Encrypt`) as well.

Spomky committed:
Refactor S/MIME encrypter to use certificate repository

Replaces direct certificate path usage with a repository interface for managing S/MIME certificates. This improves flexibility by allowing custom certificate retrieval logic through `SmimeCertificateRepositoryInterface`. Adjusted related tests, configuration, and event listener implementation accordingly.

Spomky commented:
It is important to mention in the documentation that caution should be exercised when sending encrypted emails to multiple recipients. And indeed, all recipients are used here and therefore the main recipients as well as those in the hidden folder. It is therefore possible to know these recipients depending on the encryption method used.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -2314,8 +2314,8 @@ private function addMailerSection(ArrayNodeDefinition $rootNode, callable $enabl
                             ->children()
-                                ->scalarNode('certificate')
-                                    ->info('Path to certificate (in PEM format without the `file://` prefix)')
+                                ->scalarNode('repository')
```
For reviewers, this change is fine as this feature has been introduced in 7.3, so not yet released

fabpot commented:
Thank you @Spomky.
